### ISSUE TITLE
net.Forward outputs differ from the Keras outputs

### ISSUE THREAD

**Issue by vinicius121** [opencv/opencv#16991]
##### System information (version)
- OpenCV => 4.2
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2019
- Tensorflow version => 1.15
- Keras version => 2.2.5

##### Detailed description

I have an issue concerning the classification when using net.forward. First I have trained my model in Keras with TensorFlow as backend. Then I save the model as .pb. I can import it successfully in OpenCV using 'ReadNetFromTensorflow' in C#. I'm using the OpenCVSharp nuget package which is just a wrapper. The problem I encounter is that my predictions now are not the same as I get in Keras. The accuracy differs a lot. Here is the link for the .pb [file](https://www.dropbox.com/s/iqpm6la06oegiuw/graph.pb?dl=0). Btw I was using the Flatten layer but as it is not supported by OpenCV (I was getting a error while loading the model) I have changed it to a Reshape layer as suggested by #10135. Any suggestions? Here is the code I have used for training: 

```
model = K.models.Sequential()
model.add(K.layers.Conv2D(64,kernel_size=(3,3),strides=(1, 1),input_shape=input_shape, padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.Conv2D(64,kernel_size=(3,3),strides=(1, 1), padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.MaxPooling2D(pool_size=(2,2)))

model.add(K.layers.Conv2D(128,kernel_size=(3,3),strides=(1, 1), padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.Conv2D(128,kernel_size=(3,3),strides=(1, 1), padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.MaxPooling2D(pool_size=(2,2)))

model.add(K.layers.Conv2D(256,kernel_size=(3,3),strides=(1, 1), padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.Conv2D(256,kernel_size=(3,3),strides=(1, 1), padding='same'))
model.add(K.layers.BatchNormalization())
model.add(K.layers.Activation('relu'))
model.add(K.layers.AveragePooling2D(pool_size=(2,2)))
#From Flatten to Reshape
a,b,c,d = model.output_shape
a = b*c*d
model.add(K.layers.Permute([1, 2, 3]))  # Indicate NHWC data layout
model.add(K.layers.Reshape((a,)))

model.add(K.layers.Dense(num_classes, activation='softmax', name = "output_node"))
sgd = tf.keras.optimizers.SGD(lr=lr, decay=decay, momentum=0.9, nesterov=False)

model.compile(loss=keras.losses.categorical_crossentropy, optimizer='sgd', metrics=['accuracy'])
```
Then I use the following code to obtain the .pb file:
```
sess = K.backend.get_session()
constant_graph = tf.graph_util.convert_variables_to_constants(sess, sess.graph.as_graph_def(), ['output_node/Softmax'])
tf.train.write_graph(constant_graph, "", "graph.pb", as_text=False)
```
Then, in C# I use the following code to try inference:
```
var model = System.IO.Path.Combine(Location, Model);
var net = CvDnn.ReadNetFromTensorflow(model);
var image = System.IO.Path.Combine(Location, sampleImage);
var frame = Cv2.ImRead(image);
var blob = CvDnn.BlobFromImage(frame, 1.0 / 255.0, new OpenCvSharp.Size(32, 15), new Scalar(0, 0, 0), true, false);

net.SetInput(blob);
////get output layer name
var outNames = net.GetUnconnectedOutLayersNames();
////create mats for output layer
var outs = outNames.Select(_ => new Mat()).ToArray();

using (var predictions = net.Forward(outNames[0]))
{
    PrintMat(predictions);    
}

private void PrintMat(Mat mat)
{
    for (int i = 0; i < mat.Cols; i++)
    {
        Debug.WriteLine(mat.At<float>(0, i));
    }
}
```

**Comment by dkurt** [comment]
Cannot reproduce the difference:
```python
import numpy as np
import tensorflow as tf
import cv2 as cv

np.random.seed(32)

pb_file = 'graph.pb'
graph_def = tf.compat.v1.GraphDef()

try:
    with tf.io.gfile.GFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())
except:
    with tf.gfile.FastGFile(pb_file, 'rb') as f:
        graph_def.ParseFromString(f.read())

inp = np.random.standard_normal([1, 32, 15, 3]).astype(np.float32)

with tf.Session() as sess:
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')
    tfOut = sess.run(sess.graph.get_tensor_by_name('output_node/Softmax:0'),
                     feed_dict={'conv2d_24_input:0': inp})

net = cv.dnn.readNet(pb_file)
net.setInput(inp.transpose(0, 3, 1, 2))
cvOut = net.forward()
print(np.max(np.abs(cvOut - tfOut)))
```
Please make sure that input for Keras matches input for OpenCV.

**Comment by vinicius121** [comment]
Hi! 

In Keras I do the following steps for reading the dataset and to make the prediction:
Here is the keras [model](https://www.dropbox.com/s/we2zuzu61h63c0s/kerasTF.h5?dl=0)
```
image =cv2.imread(datapath + filename + '.png')
image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB) 
data.append(image)
#shape is (32,15,3)
#make prediction
model.predict(np.expand_dims(data_x[0]/255, axis=0))
```
In python I'm doing the following:
```
inp = cv.imread('Data/' + 'image' + '.png')
blob = cv.dnn.blobFromImage(inp, scalefactor=1.0/255, mean=(0, 0, 0),size=(15, 32), swapRB=True)
#shape (1,3,32,15)
net.setInput(blob) 
cvOut = net.forward()
```
Am I missing something?

**Comment by dkurt** [comment]
@vinicius121, looks like `.pb` and `.h5` just are different models. Please try to convert `.h5` to `.pb` once again using this script:

```python
import tensorflow as tf
import numpy as np
import cv2 as cv
from tensorflow import keras as K
print(K.__version__)

# Convert Keras .h5 to TensorFlow .pb
model = K.models.load_model('kerasTF.h5')
sess = K.backend.get_session()
sess.as_default()

from tensorflow.python.tools import optimize_for_inference_lib
from tensorflow.tools.graph_transforms import TransformGraph
graph_def = sess.graph.as_graph_def()
graph_def = tf.graph_util.convert_variables_to_constants(sess, graph_def, ['output_node/Softmax'])

with tf.gfile.FastGFile('kerasTF.pb', 'wb') as f:
   f.write(graph_def.SerializeToString())

# Run Keras
image = cv.imread('/home/dkurt/Pictures/Untitled.png')

kerasInp = cv.resize(image, (15, 32))
kerasInp = cv.cvtColor(kerasInp, cv.COLOR_BGR2RGB)
data = []
data.append(kerasInp)
kerasOut = model.predict(np.expand_dims(data[0]/255, axis=0))

# Run OpenCV
net = cv.dnn.readNet('kerasTF.pb')
blob = cv.dnn.blobFromImage(image, scalefactor=1.0/255, mean=(0, 0, 0),size=(15, 32), swapRB=True)
net.setInput(blob)
cvOut = net.forward()

# Compare
print(cvOut.shape)
print(kerasOut.shape)
print(np.max(np.abs(cvOut - kerasOut)))
```
output for me:
```
4.48199e-09
```
It I compare the same Keras model with a `graph.pb` from issue description - output will be different. But with `.pb` from this `.h5` - they matches.

**Comment by vinicius121** [comment]
Hi!
I redid everything again. [Here ](https://www.dropbox.com/s/v6i1wiepxr4u4wb/Tests.zip?dl=0)it is available the model in both .h5 and .pb (same model in both files).
I did some tests and I still get different values. So in the same link there is the script that I'm using as well as some sample images. In the script what I do is 2 tests:
-The first I'm reading the image with OpenCV and then I'm turning it as a numpy array.
-The second I'm also reading the image with OpenCV but then I use the blob function.
There is also a .txt file with my results.
Thanks

**Comment by dkurt** [comment]
Have you tried to run ode sample from https://github.com/opencv/opencv/issues/16991#issuecomment-609467748? It looks like a bug in your application, but not OpenCV.

**Comment by vinicius121** [comment]
Hi, yes I have. But in your script you resize the image before using Keras. That's why I have shared a test script in my previous comment. Probably it is a bug in my application. However, the script that I shared [here ](https://www.dropbox.com/s/n2a7qxgy9hki4sh/Tests.zip?dl=0) is a simple one, the only change is how do I import a sample image. And it gives different results using the same model with OpenCV. Could you try to run the script please? Maybe I'm missing something.

**Comment by dkurt** [comment]
@vinicius121, do you know that images from `Data` folder have 32 pixel width and 15 pixel height but the model is trained on 1x32x15x3 input tensors that means 32 height and 15 width. Is that right?

**Comment by dkurt** [comment]
Please if you have a usage question use a forum instead: http://answers.opencv.org/. The model works correctly and it's obviously a bug in the application.

**Comment by vinicius121** [comment]
Hi! Yes I know. That's why I change the axis to be the same using Numpy and I get the same predictions as in Keras. And you are correct it is not a bug, and the model works. My confusion is with the blob function. And yes I have tried the [forum](https://answers.opencv.org/question/228468/netforward-outputs-differ-from-the-keras-outputs/). However, I got no answer there. I will explore better.
Thank you for your help!
